File: go/iban.go
package main

// isValidIban checks an IBAN per ISO 13616: structure plus the ISO 7064 MOD-97-10 check.
func isValidIban(iban string) bool {
	if len(iban) < 15 || len(iban) > 34 {
		return false
	}

	for i := 0; i < len(iban); i++ {
		c := iban[i]
		var ok bool
		switch {
		case i < 2:
			ok = isUpper(c) // country code
		case i < 4:
			ok = isDigit(c) // check digits
		default:
			ok = isUpper(c) || isDigit(c)
		}
		if !ok {
			return false
		}
	}

	// Move the first four characters to the end, map A=10..Z=35, then take mod 97.
	// The rearranged number is far too large for any integer type, so the remainder
	// is folded in digit by digit.
	remainder := 0
	for k := 0; k < len(iban); k++ {
		c := iban[(k+4)%len(iban)]
		if isDigit(c) {
			remainder = (remainder*10 + int(c-'0')) % 97
		} else {
			remainder = (remainder*100 + int(c-'A') + 10) % 97
		}
	}

	return remainder == 1
}

// bankCode returns the four-digit bank code: characters 5-8 of the IBAN.
func bankCode(iban string) string { return iban[4:8] }

File: cpp/src/iban.cpp
#include "iban.h"

#include <cstddef>

namespace settle {
namespace {

bool isDigit(char c) { return c >= '0' && c <= '9'; }
bool isUpper(char c) { return c >= 'A' && c <= 'Z'; }

}  // namespace

bool isValidIban(std::string_view iban) {
    if (iban.size() < 15 || iban.size() > 34) return false;

    for (std::size_t i = 0; i < iban.size(); ++i) {
        const char c = iban[i];
        const bool ok = i < 2   ? isUpper(c)                 // country code
                        : i < 4 ? isDigit(c)                 // check digits
                                : isUpper(c) || isDigit(c);
        if (!ok) return false;
    }

    // Move the first four characters to the end, map A=10..Z=35, then take mod 97.
    // The rearranged number is far too large for any integer type, so the remainder
    // is folded in digit by digit.
    int remainder = 0;
    for (std::size_t k = 0; k < iban.size(); ++k) {
        const char c = iban[(k + 4) % iban.size()];
        remainder = isDigit(c) ? (remainder * 10 + (c - '0')) % 97
                               : (remainder * 100 + (c - 'A' + 10)) % 97;
    }

    return remainder == 1;
}

std::string_view bankCode(std::string_view iban) { return iban.substr(4, 4); }

}  // namespace settle
